File: swift/416_PartitionEqualSubsetSum.swift
class Solution {
    func canPartition(_ nums: [Int]) -> Bool {
        var total = 0
        var N = nums.count
        for n in nums {
            total += n
        }
        if total % 2 != 0 { return false } 
        
        total /= 2
        var dp = [[Bool]](repeating: [Bool](repeating: false, count: total + 1), count: N + 1)
        dp[0][0] = true
        
        for i in 1...N {
            var currentNumber = nums[i - 1]
            for j in 0...total {
                dp[i][j] = dp[i - 1][j] || (j >= currentNumber && dp[i - 1][j - currentNumber])
            }
        }
        return dp[N][total]
    }
}

File: cpp/416_PartitionEqualSubsetSum.cpp
class Solution {
public:
    unordered_map<int, bool> cache;
    bool canPartition(vector<int> &nums) {
        int totalSum = 0;
        // find sum of all array elements
        for (int num : nums) {
            totalSum += num;
        }
        // if totalSum is odd,it cannot be partitioned into equal sum subset
        if (totalSum % 2 != 0) return false;
        int subSetSum = totalSum / 2;
        int n = nums.size();
        return dfs(nums, n - 1, subSetSum);
    }

    bool dfs(vector<int> &nums, int n, int subSetSum) {
        // Base Cases
        if (cache.find(subSetSum) != cache.end()) 
            return cache[subSetSum];
        if (subSetSum == 0)
            return true;
        if (n == 0 || subSetSum < 0)
            return false;
        
        cache[subSetSum] = dfs(nums, n - 1, subSetSum - nums[n - 1]) || dfs(nums, n - 1, subSetSum);
        return cache[subSetSum];
    }
};

